### ISSUE TITLE
DnsNameResolverTest#testNonCachedResolve fails for localhost in IPv6 only network

### ISSUE THREAD

**Issue by slandelle** [netty/netty#6281]
### Expected behavior

DnsNameResolverTest#testNonCachedResolve succeeds.

### Actual behavior

`DnsNameResolverTest#testNonCachedResolve` fails because it wants IPv4:

```java
DnsNameResolver resolver = newNonCachedResolver(InternetProtocolFamily2.IPv4).build();
```

```
unresolved=localhost
resolved.hostAddress=0:0:0:0:0:0:0:1
resolved.class=class java.net.Inet6Address
resolver.resolvedAddressTypes=class java.net.Inet4Address
```

You can check [AsyncHttpClient's CI on Travis](https://travis-ci.org/AsyncHttpClient/async-http-client/builds/195491450#L1696).
Note that test code lines don't exactly match because I added some more logs as the `assertThat` in `testResolve0` don't provide any reason.

### Steps to reproduce

Run test on IPv6 only network, such as Travis.

### Netty version

4.1 master

### JVM version (e.g. `java -version`)

```
java version "1.8.0_121"
Java(TM) SE Runtime Environment (build 1.8.0_121-b13)
Java HotSpot(TM) 64-Bit Server VM (build 25.121-b13, mixed mode)
```

### OS version (e.g. `uname -a`)

Debian

**Comment by normanmaurer** [comment]
@slandelle actually this is expected as you use `InternetProtocolFamily.IPv4`. Maybe we should add some assumeTrue to skip tests that need ipv4 while only ipv6 is present ?

**Comment by slandelle** [comment]
Check that `NetUtil.LOCALHOST4 != null`?

**Comment by normanmaurer** [comment]
@slandelle need to think about it... this will not work.

**Comment by normanmaurer** [comment]
@slandelle so the real problem here is that the HostFileParser will not correctly handle the case where the hosts file contain an entry for ipv4 and ipv6. It will only preserve the first found which is incorrect as it valid to have one for each. 

Thinking about a way to fix this without api breakage ...

**Comment by slandelle** [comment]
Ouch, I missed that :(

Do you want me to help on this?
Shouldn't `HostsFileParser` be annotated `@UnstableApi`?

**Comment by normanmaurer** [comment]
@slandelle yes it should but we missed it so we can not easily breaking it...

**Comment by slandelle** [comment]
We could leave existing methods as is and deprecate them, and add new methods that would return a pair of Maps, one for IPv4 and one for IPv6. WDYT?

**Comment by normanmaurer** [comment]
@slandelle we can not add methods to interfaces that are public in a bug-fix release.

**Comment by slandelle** [comment]
@normanmaurer Interface? `HostsFileParser` is a final class, so no one can extend it without seriously hacking around.

The only way I see adding new methods there would be harmful is someone having some algorithm based on the number and order of methods fetched from reflection, and I fail to see why someone would do that. Am I missing something?

**Comment by normanmaurer** [comment]
@slandelle I am talking about the `HostNameResolver` that uses it here.

**Comment by slandelle** [comment]
@normanmaurer You probably mean `HostsFileEntriesResolver`. Yeah, got it.

We could introduce a new fixed interface with the new behavior, have `DefaultHostsFileEntriesResolver`, check type once in `DnsNameResolver` and downcast.

**Comment by normanmaurer** [comment]
@slandelle yeah maybe something like this... want to do a pr ?

**Comment by slandelle** [comment]
@normanmaurer Yep. Can't it wait for a few days? Release day on my side...

**Comment by normanmaurer** [comment]
sure

**Comment by slandelle** [comment]
great! stay tuned :)

**Comment by slandelle** [comment]
@normanmaurer Is it possible to make the `resolver` module depend on the `transport` one?

I'd like to introduce a new `InetAddress address(String inetHost, InternetProtocolFamily[] resolvedAddressTypes)` on `DefaultHostsFileEntriesResolver` but `InternetProtocolFamily` is currently not reachable :(

**Comment by slandelle** [comment]
I have the feeling `InternetProtocolFamily` should sit in `commons` side to `NetUtils` :(

**Comment by normanmaurer** [comment]
@slandelle let me think about it... one thing is for sure, we can not move it :(

**Comment by slandelle** [comment]
I know :)
Anyway, it feels natural that resolving an address against a hosts file would take a parameter indicating which `InternetProtocolFamily`s the user wants/prefers.

IMO, it also means that `DnsNameResolver::DEFAULT_RESOLVE_ADDRESS_TYPES` should be a public API. FIY, in order to not duplicate this logic, I created a class in the `io.netty.resolver.dns` package that exposes it publicly. I was thinking about introducing an enum (4 values) so we wouldn't expose a mutable array.

**Comment by Scottmitch** [comment]
IIRC the intention is that the entire dns package is unstable. I don't think we should invest much effort in preserving APIs if they are broken. If we forgot to put `@UnstableApi`in a PR lets fix the API and add this annotation.

**Comment by slandelle** [comment]
👍 I fail to see why someone would expect this class in particular to be stable while all the others in the same module are.

**Comment by normanmaurer** [comment]
@slandelle @Scottmitch actually it's in the resolver module so not sure it's true tho

**Comment by Scottmitch** [comment]
the `resolver` module doesn't exist in `4.0` ... are you saying this module was intentionally not marked as `UnstableApi`? Much of this code was introduced late in a candidate release for `4.1` [1] ... I think it is more likely we just forgot to annotate the package, no?

[1] https://github.com/netty/netty/commit/8d4db050f36929430e6699eec640de8eab1f3df2

**Comment by normanmaurer** [comment]
Yes we most likely forgot that said I think we can not just do whatever we want now ;) 

That said after a bit more of thoughts I think change from InetAddress to List<InetAdress> in this case should not affect people. At least chances are very small, so @slandelle go for it

**Comment by slandelle** [comment]
@normanmaurer Actually, what I had in mind was to still only return one single address, but add an array of protocol families array to tell the resolver which is preferred. WDYT?

**Comment by Scottmitch** [comment]
> Yes we most likely forgot that said I think we can not just do whatever we want now ;)

I understand however the `@UnstableApi` annotations are a nice indicator but they are easy to forget and are not foolproof. I don't think just because we forgot these annotations we should bind ourselves to this API being stable. My preference is to annotate these classes with `@UnstableApi` (as originally intended) if we can't find a good solution.

**Comment by normanmaurer** [comment]
@slandelle sounds good.

**Comment by slandelle** [comment]
`transport` depends on `resolver` so I can't of course have the other way around 😭 😭 😭

**Comment by slandelle** [comment]
PTAL #6334

**Comment by slandelle** [comment]
@normanmaurer This issue can be closed since #6334 was merged.
